nni/compression/pytorch/speedup/compress_modules.py

```diff
@@ -531,13 +564,90 @@ def replace_layernorm(layernorm, masks):
     if len(in_masks) != 1:
         raise InputsNumberError()
     in_mask = in_masks[0]
-    dim_n = len(in_mask.size())
-    new_shape = []
-    for i in range(1, dim_n):
-        sum_dims = list(range(0, dim_n))
-        sum_dims.remove(i)
-        reduced = torch.sum(in_mask, sum_dims)
-        n_remained = torch.sum(reduced > 0)
-        new_shape.append(n_remained)
-
-    return nn.LayerNorm(tuple(new_shape), layernorm.eps, layernorm.elementwise_affine)
+    dense_shape = convert_dense_shape(in_mask)
+    norm_shape = layernorm.normalized_shape
+    dim_n = len(dense_shape) - len(norm_shape)
+    return nn.LayerNorm(dense_shape[dim_n:], layernorm.eps, layernorm.elementwise_affine)
+
+
+def replace_embedding(embedding, masks):
+    """
+    Replace the embedding layer according the infered masks.
+    We replace the embedding layer according the weight masks,
+    """
+    # currently we donnot support replace the embedding layer
+    # because we donnot have the corressponding pruner
+    return embedding
+    # in_masks, out_masks, weight_masks = masks
+    # w_mask = weight_masks['weight']
+    # # weight size [num_embeddings, embeddings_dim]
+    # # currently, only support the pruning on the embedding_dim
+    # # dimension.
+    # pruned_dim, remained_dim = convert_to_coarse_mask(w_mask, 1)
+
+    # n_remain_dim = remained_dim.size(0)
+    # num_embeddings = embedding.num_embeddings
+    # padding_idx = embedding.padding_idx
+    # max_norm = embedding.max_norm
+    # norm_type = embedding.norm_type
+    # scale_grad_by_freq = embedding.scale_grad_by_freq
+    # new_embedding = torch.nn.Embedding(num_embeddings, n_remain_dim, padding_idx=padding_idx,
+    #                                    max_norm=max_norm, norm_type=norm_type, scale_grad_by_freq=scale_grad_by_freq)
+    # new_embedding.weight.data = torch.index_select(embedding.weight.data, 1, remained_dim)
+    # return new_embedding
+
+
+def replace_view(trace_model, cpp_node, masks):
```

Comment:
Do we already fix the view issue in our last discussion?

Reply:
Nice catch, I forgot this one. Thanks.